optuna/integration/mlflow.py

```diff
@@ -0,0 +1,141 @@
+import optuna
+from optuna import type_checking
+
+if type_checking.TYPE_CHECKING:
+    from typing import Dict  # NOQA
+
+try:
+    import mlflow
+
+    _available = True
+except ImportError as e:
+    _import_error = e
+    _available = False
+    mlflow = object
+
+
+def _check_mlflow_availability():
+    # type: () -> None
+
+    if not _available:
+        raise ImportError(
+            "MLflow is not available. Please install MLflow to use this "
+            "feature. It can be installed by executing `$ pip install "
+            "mlflow`. For further information, please refer to the installation guide "
+            "of MLflow. (The actual import error is as follows: " + str(_import_error) + ")"
+        )
+
+
+class MLflowCallback(object):
+    """Callback to track optuna trials with MLflow.
+
+    This callback adds relevant information that is
+    tracked by Optuna to MLflow.
+
+    Example:
+
+        Add MLflow callback to optuna optimization.
+
+        .. testsetup::
+
+            import pathlib
+            import tempfile
+            tempdir = tempfile.mkdtemp()
+            YOUR_TRACKING_URI = pathlib.Path(tempdir).as_uri()
+
+        .. testcode::
+
+            import optuna
+            from optuna.integration.mlflow import MLflowCallback
+
+            def objective(trial):
+                x = trial.suggest_uniform('x', -10, 10)
+                return (x - 2) ** 2
+
+            mlflc = MLflowCallback(
+                tracking_uri=YOUR_TRACKING_URI,
+                metric_name='my metric score',
+            )
+
+            study = optuna.create_study(study_name='my_study')
+            study.optimize(objective, n_trials=10, callbacks=[mlflc])
+
+        .. testcleanup::
+
+            import shutil
+            shutil.rmtree(tempdir)
+
+        .. testoutput::
+            :hide:
+            :options: +NORMALIZE_WHITESPACE
+
+            INFO: 'my_study' does not exist. Creating a new experiment
+
+    Args:
+        tracking_uri:
+            Set the tracking server URI.
+
+            - An empty string, or a local directory path, prefixed with ``file:``. Data is stored
+              locally in the provided directory (or ``./mlruns`` if empty).
+            - An HTTP URI like ``https://my-tracking-server:5000``.
+            - A Databricks workspace, provided as the string ``databricks`` or, to use a
+              Databricks CLI.
+              `profile <https://github.com/databricks/databricks-cli#installation>`_,
+              ``databricks://<profileName>``.
+        experiment:
+            Name of MLflow experiment to be activated. If not set ``study.study_name``
+            will be taken. Either ``experiment`` or ``study.study_name`` must be set.
+        metric_name:
+            Name of the metric. If not provided this will be called ``trial_value``.
+    """
+
+    def __init__(self, tracking_uri=None, experiment=None, metric_name=None):
```

Comment:
Just curious but in what cases would you like to specify a different experiment name than the one implied by the study (name)?

Reply:
Since Optuna does not force the user to provide a study name I do not want to force the user to provide one when he is using this callback. So I introduced the experiment name parameter and enforce the user to provide either this param or a study name.

@hvy Do you think I should enforce the user to provide a study name?

Reply:
I see, thanks for clarifying your thoughts on this. I think that if the user can specify the experiment name, he/she could as well specify the study name, and those would be consistent. I think this parameter makes it look like they should be treated independently (which I now understand is not the case). How about omitting this parameter and writing in the docs how the experiment name is determined?

Reply:
@hvy What do you suggest happens when the user does not provide a study name? Raise an exception? I do not like to use the default experiment name because that is the place where "garbage" is collected when the user forgets to provide a study name.

Reply:
> that is the place where "garbage" is collected

Hmm, I’m not sure if this is a general assumption. 

Even if it is the case, we can later change the behavior depending on users' feedback, since this is an `experimental` feature. Thus, I think we should adopt the simpler interface so far, where we just omit the parameter as @hvy mentions and just take the study name.

Reply:
It might help if you could provide some concrete scenario. It sounds to me like if the user would forget to specify the study name (--> "garbage"), he/she is most likely to forget specifying the experiment name as well. Additionally, having to recall that there are two interfaces to specify the experiment name incurs additional load. I otherwise agree with @g-votte what we should start simple and expand the API as needed.

Reply:
Ok. I made a suggestion. Now the MLflow expermiment name is taken from the Optuna study_name and from nowhere else. But I enforce the study_name to be set or an exception is raised.

What do you think?

Reply:
Thanks @PhilipMay for your changes. On the exception however, I don't think raising one here speaks for good design. First, I'm not sure the general user would find that intuitive, and second the name is currently hardcoded in the MLflow integration module and will only work for the in memory storage and seems fragile.

Reply:
Changed it to just use study_name.